File: src/main/java/com/refinedmods/refinedstorage/RS.java
package com.refinedmods.refinedstorage;

import com.refinedmods.refinedstorage.apiimpl.API;
import com.refinedmods.refinedstorage.config.ClientConfig;
import com.refinedmods.refinedstorage.config.ServerConfig;
import com.refinedmods.refinedstorage.datageneration.DataGenerators;
import com.refinedmods.refinedstorage.integration.curios.CuriosIntegration;
import com.refinedmods.refinedstorage.network.NetworkHandler;
import com.refinedmods.refinedstorage.setup.ClientSetup;
import com.refinedmods.refinedstorage.setup.CommonSetup;
import com.refinedmods.refinedstorage.setup.ServerSetup;

import net.neoforged.api.distmarker.Dist;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.fml.ModLoadingContext;
import net.neoforged.fml.common.Mod;
import net.neoforged.fml.config.ModConfig;
import net.neoforged.fml.loading.FMLEnvironment;
import net.neoforged.neoforge.common.NeoForge;

@Mod(RS.ID)
public final class RS {
    public static final String ID = "refinedstorage";
    public static final String NAME = "Refined Storage";

    public static final NetworkHandler NETWORK_HANDLER = new NetworkHandler();
    public static final ServerConfig SERVER_CONFIG = new ServerConfig();
    public static final ClientConfig CLIENT_CONFIG = new ClientConfig();

    public RS(IEventBus eventBus) {
        if (FMLEnvironment.dist == Dist.CLIENT) {
            eventBus.addListener(ClientSetup::onClientSetup);
            eventBus.addListener(ClientSetup::registerMenuScreens);
            eventBus.addListener(ClientSetup::onModelBake);
            eventBus.addListener(ClientSetup::onRegisterAdditionalModels);
            eventBus.addListener(ClientSetup::onRegisterModelGeometry);
            eventBus.addListener(ClientSetup::onRegisterKeymappings);
            eventBus.addListener(ClientSetup::onRegisterColorBindings);
            NeoForge.EVENT_BUS.addListener(ClientSetup::addReloadListener);
        }

        NeoForge.EVENT_BUS.register(new ServerSetup());

        ModLoadingContext.get().registerConfig(ModConfig.Type.SERVER, SERVER_CONFIG.getSpec());
        ModLoadingContext.get().registerConfig(ModConfig.Type.CLIENT, CLIENT_CONFIG.getSpec());

        RSBlocks.register(eventBus);
        RSItems.register(eventBus);

        eventBus.addListener(CommonSetup::onCommonSetup);
        eventBus.addListener(CommonSetup::onRegister);
        eventBus.addListener(CommonSetup::onRegisterCapabilities);
        eventBus.addListener(CommonSetup::onRegisterNetworkPackets);
        eventBus.register(new DataGenerators());
        eventBus.register(new CuriosIntegration());

        RSContainerMenus.REGISTRY.register(eventBus);
        RSBlockEntities.REGISTRY.register(eventBus);
        RSRecipeSerializers.REGISTRY.register(eventBus);

        API.deliver();
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/RS.kt
package com.refinedmods.refinedstorage

//import com.refinedmods.refinedstorage.config.ClientConfig
//import com.refinedmods.refinedstorage.config.ServerConfig
//import com.refinedmods.refinedstorage.network.NetworkHandler
import com.refinedmods.refinedstorage.extensions.getCustomLogger
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.client.itemgroup.FabricItemGroupBuilder
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemStack
import net.minecraft.util.Identifier

object RS: ModInitializer {
    val log = getCustomLogger(RS::class)
    const val ID = "refinedstorage"
//    val NETWORK_HANDLER = NetworkHandler()
//    val MAIN_GROUP: ItemGroup = FabricItemGroupBuilder
//            .create(Identifier(ID, ID))
//            .icon {ItemStack(RSBlocks.CREATIVE_CONTROLLER) }
//            .build() // TODO add to item group
//    val SERVER_CONFIG = ServerConfig()
//    val CLIENT_CONFIG = ClientConfig()

    override fun onInitialize() {
        // TODO Register stuff!
        log.info("Initializing...")
//        DistExecutor.safeRunWhenOn(Dist.CLIENT, { { ClientSetup() } })
//        MinecraftForge.EVENT_BUS.register(ServerSetup())
//        ModLoadingContext.get().registerConfig(ModConfig.Type.SERVER, SERVER_CONFIG.getSpec())
//        ModLoadingContext.get().registerConfig(ModConfig.Type.CLIENT, CLIENT_CONFIG.getSpec())
//        val commonSetup = CommonSetup()
//        FMLJavaModLoadingContext.get().getModEventBus().addListener({ e: ? -> commonSetup.onCommonSetup(e) })
//        FMLJavaModLoadingContext.get().getModEventBus().addGenericListener(Block::class.java, { e:<net.minecraft.block.Block?> -> commonSetup.onRegisterBlocks(e) })
//        FMLJavaModLoadingContext.get().getModEventBus().addGenericListener(BlockEntityType::class.java, { e:<<>?> -> commonSetup.onRegisterTiles(e) })
//        FMLJavaModLoadingContext.get().getModEventBus().addGenericListener(Item::class.java, { e:<net.minecraft.item.Item?> -> commonSetup.onRegisterItems(e) })
//        FMLJavaModLoadingContext.get().getModEventBus().addGenericListener(IRecipeSerializer::class.java, { e:<<>?> -> commonSetup.onRegisterRecipeSerializers(e) })
//        FMLJavaModLoadingContext.get().getModEventBus().addGenericListener(ContainerType::class.java, { e:<<>?> -> commonSetup.onRegisterContainers(e) })
//        deliver()
        log.info("Initialized!")
    }
}